feat(plugin-iceberg): Add $snapshot_sequence_number as hidden column in iceberg table

agrawalreetika opened the pull request:
## Description
Add `$snapshot_sequence_number` as a hidden column in the iceberg table

## Motivation and Context
- Add `$snapshot_sequence_number` extra column in `$snapshots` metadata table
- Add `$snapshot_sequence_number` as a hidden column in the iceberg table

As `$snapshot_id` is not incremental so while calculating Delta between 2 snapshots (With query like `WHERE $snapshot_id BETWEEN snap1 AND snap2`) would return wrong results since comparison is not valid.

## Impact
Get the delta between 2 snapshots using `$snapshot_sequence_number` as column.

Here these are the covered options -
1. `WHERE "$snapshot_sequence_number" >= X` (Get delta from start snapshot to X)
2. `WHERE "$snapshot_sequence_number" BETWEEN X AND Y` (Get delta from start snapshot to X)
3. `WHERE "$snapshot_sequence_number" = X`
4. `WHERE "$snapshot_sequence_number" > X`

**Limitation -** 
Above are using `newIncrementalAppendScan()` Iceberg API which is append-only snapshot operation mode. Also `IncrementalChangelogScan` has limitation and doesn't support delta for snapshots with operation modes `delete` & `overwrite` and fails [here](https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/BaseIncrementalChangelogScan.java#L110).

## Test Plan
Tests Added

- Get `$snapshot_sequence_number `from `$snapshots` metadata table
- Now we can use `$snapshot_sequence_number `column for filter to find delta on a table

## Contributor checklist

- [ ] Please make sure your submission complies with our [contributing guide](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md), in particular [code style](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#code-style) and [commit standards](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#commit-standards).
- [ ] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.
- [ ] If adding new dependencies, verified they have an [OpenSSF Scorecard](https://securityscorecards.dev/#the-checks) score of 5.0 or higher (or obtained explicit TSC approval for lower scores).

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

Iceberg Connector Changes
* Add ``$snapshot_sequence_number`` as a hidden column in the Iceberg table.
```

hantangwangd commented:
@agrawalreetika thanks for this change. A quick question: can you help me understand the decision behind using an inclusive lower bound for the incremental query rather than an exclusive one? For our use cases like refreshing materialized views, wouldn't the latter be more appropriate?

hantangwangd commented:
>  For `WHERE $snapshot_sequence_number >= X`, we should take the table state from start to X (So used inclusive)

Thanks for the explanation, I'm trying to make sure I understand the filter semantics correctly. As I understand it, `WHERE $snapshot_sequence_number >= X` means `getting the incremental data of the table from X-1 to the newest,` is this correct?

If that's the case, what do you think about supporting and using `WHERE $snapshot_sequence_number > X`? It seems to map more directly to `getting the incremental data of the table from X to the newest`. What do you think? Please correct me if I've misunderstood anything.

agrawalreetika commented:
> > For `WHERE $snapshot_sequence_number >= X`, we should take the table state from start to X (So used inclusive)
> 
> Thanks for the explanation, I'm trying to make sure I understand the filter semantics correctly. As I understand it, `WHERE $snapshot_sequence_number >= X` means `getting the incremental data of the table from X-1 to the newest,` is this correct?
> 
> If that's the case, what do you think about supporting and using `WHERE $snapshot_sequence_number > X`? It seems to map more directly to `getting the incremental data of the table from X to the newest`. What do you think? Please correct me if I've misunderstood anything.

`WHERE $snapshot_sequence_number >= X` - @hantangwangd, you are right about this, I think committed different condition changes, updated it now. Also covered  `WHERE $snapshot_sequence_number > X` - Where we take X+1 to the latest delta.

I have updated description with the covered scenario, please take a look and lmk if anything is missing.

PingLiuPing commented:
I agree with @hantangwangd comment's. Especially on 3. I wonder the semantic correctness of `WHERE "$snapshot_sequence_number" = X`. From iceberg spec,  I think it is used to track the relative “age” of data and delete files. From Iceberg source code I think 4 is correct.

hantangwangd commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/changelog/ChangelogSplitSource.java:
```diff
@@ -78,7 +78,8 @@ public ChangelogSplitSource(
             Table table,
-            IncrementalChangelogScan tableScan)
+            IncrementalChangelogScan tableScan,
+            long snapshotSequenceNumber)
```
```suggestion
            IncrementalChangelogScan tableScan)
```

nit: missed cleaning this?

hantangwangd commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergSplitManager.java:
```diff
@@ -93,7 +119,7 @@ public ConnectorSplitSource getSplits(
                     .fromSnapshotExclusive(fromSnapshot)
                     .toSnapshot(toSnapshot);
-            return new ChangelogSplitSource(session, typeManager, icebergTable, scan);
+            return new ChangelogSplitSource(session, typeManager, icebergTable, scan, icebergTable.snapshot(toSnapshot).sequenceNumber());
```
```suggestion
            return new ChangelogSplitSource(session, typeManager, icebergTable, scan);
```

nit: the sequence number seems unnecessary here

hantangwangd commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergSplit.java:
```diff
@@ -143,7 +143,6 @@ public NodeSelectionStrategy getNodeSelectionStrategy()
     {
         return nodeSelectionStrategy;
     }
-
```
nit: revert this change?

hantangwangd commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergSplitSource.java:
```diff
@@ -83,6 +84,24 @@ public IcebergSplitSource(
+    public IcebergSplitSource(
+            ConnectorSession session,
+            IncrementalAppendScan scan,
+            TupleDomain<IcebergColumnHandle> metadataColumnConstraints)
```
```suggestion
    public <T> IcebergSplitSource(
            ConnectorSession session,
            Scan<T, FileScanTask, CombinedScanTask> tableScan,
            TupleDomain<IcebergColumnHandle> metadataColumnConstraints)
```

I wonder if we can use a unified constructor like this for both `TableScan` and `IncrementalAppendScan`?

hantangwangd commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergUtil.java:
```diff
@@ -494,6 +497,73 @@ public static Map<String, Integer> columnNameToPositionInSchema(Schema schema)
+        long lowerSequenceNumber = ((Number) range.getLowBoundedValue()).longValue();
+        long upperSequenceNumber = ((Number) range.getHighBoundedValue()).longValue();
+
+        Snapshot lower = getSnapshotBySequenceNumber(table, lowerSequenceNumber);
```
Is this intentional not to validate?

hantangwangd commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergUtil.java:
```diff
@@ -494,6 +497,73 @@ public static Map<String, Integer> columnNameToPositionInSchema(Schema schema)
+        Snapshot upper = validateAndFetchEndSnapshot(table, upperSequenceNumber);
+        hasOnlyAppendOperations(table, lower, upper);
+
+        return new IcebergTableName(name.getTableName(), INCREMENTAL, Optional.of(lower.snapshotId()), Optional.of(upper.snapshotId()));
```
nit: the `lower` might be null here.

agrawalreetika replied:
Handled in validateAndFetchSnapshot

hantangwangd commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergAbstractMetadata.java:
```diff
@@ -351,9 +361,62 @@ public ConnectorTableLayoutResult getTableLayoutForConstraint(
+            }
+        }
+
+        TupleDomain<ColumnHandle> newDomainPredicate = TupleDomain.withColumnDomains(newDomains.build());
```
```suggestion
        TupleDomain<ColumnHandle> newDomainPredicate;
        Optional<Map<ColumnHandle, Domain>> optionalDomains = constraint.getSummary().getDomains();
        if (optionalDomains.isPresent()) {
            Map<ColumnHandle, Domain> domains = optionalDomains.get();
            ImmutableMap.Builder<ColumnHandle, Domain> newDomains = ImmutableMap.builder();
            for (Map.Entry<ColumnHandle, Domain> entry : domains.entrySet()) {
                IcebergColumnHandle column = (IcebergColumnHandle) entry.getKey();

                if (!column.getName().equalsIgnoreCase("$snapshot_sequence_number")) {
                    newDomains.put(entry.getKey(), entry.getValue());
                    continue;
                }

                Domain domain = entry.getValue();
                if (domain.isSingleValue()) {
                    long sequenceNumber = ((Number) domain.getSingleValue()).longValue();
                    Snapshot endSnapshot = validateAndFetchSnapshot(icebergTable, sequenceNumber);
                    handle = handle.withUpdatedIcebergTableName(prepareIncrementalScan(icebergTable, endSnapshot, name, endSnapshot, true), true);
                    continue;
                }

                List<Range> ranges = domain.getValues().getRanges().getOrderedRanges();
                if (ranges.size() != 1) {
                    throw unsupportedPredicate();
                }

                Range range = ranges.get(0);
                if (range.isSingleValue()) {
                    long seqNum = ((Number) range.getSingleValue()).longValue();
                    Snapshot endSnapshot = validateAndFetchSnapshot(icebergTable, seqNum);
                    handle = handle.withUpdatedIcebergTableName(prepareIncrementalScan(icebergTable, endSnapshot, name, endSnapshot, true), true);
                }
                else if (!range.isLowUnbounded() && range.isLowInclusive() && range.isHighUnbounded()) {
                    // WHERE $snapshot_sequence_number >= X (delta from x to latest)
                    long lowerSequenceNumber = ((Number) range.getLowBoundedValue()).longValue();
                    Snapshot start = validateAndFetchSnapshot(icebergTable, lowerSequenceNumber);
                    handle = handle.withUpdatedIcebergTableName(prepareIncrementalScan(icebergTable, start, name, latestSnapshot, true), true);
                }
                else if (!range.isLowUnbounded() && !range.isLowInclusive() && range.isHighUnbounded()) {
                    // WHERE $snapshot_sequence_number > X, excludes lower bound
                    long lowerSequenceNumber = ((Number) range.getLowBoundedValue()).longValue();
                    Snapshot start = validateAndFetchSnapshot(icebergTable, lowerSequenceNumber);
                    handle = handle.withUpdatedIcebergTableName(prepareIncrementalScan(icebergTable, start, name, latestSnapshot, false), false);
                }
                else if (range.isLowInclusive() && range.isHighInclusive()) {
                    // WHERE $snapshot_sequence_number BETWEEN X AND Y
                    handle = handle.withUpdatedIcebergTableName(prepareBetweenScan(icebergTable, name, range, false), false);
                }
                else {
                    throw unsupportedPredicate();
                }
            }
            newDomainPredicate = TupleDomain.withColumnDomains(newDomains.build());
        }
        else {
            newDomainPredicate = TupleDomain.none();
        }

```

After PR #26374, we need to account for the case where the original domain predicate is `none`. The CI test failure appears to be due to this.

agrawalreetika replied:
Thanks for pointing it out @hantangwangd, I have updated the changes.

tdcmeehan commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergUtil.java:
```diff
@@ -494,6 +497,157 @@ public static Map<String, Integer> columnNameToPositionInSchema(Schema schema)
+            }
+            else if (range.isLowInclusive() && range.isHighInclusive()) {
+                // WHERE $snapshot_sequence_number BETWEEN X AND Y
+                handle = handle.withUpdatedIcebergTableName(prepareBetweenScan(icebergTable, name, range, false), false);
```
SQL between is inclusive, so shouldn't the fromInclusive flag be set to true?
